### PR TITLE
Sharded matrix multiplication and tests refactor

### PR DESCRIPTION
**Pull request by lyprince**
Sharded matrix multiplications stored in the `sharded` submodule. Only three were implemented to support common cases for tensor and data parallel training applications.

Tests for collectives were refactored to reuse utilities.

### CODE REVIEW

File: poptorch_experimental_addons/_impl/collectives.py [hunk 1]
```diff
@@ -39,6 +42,20 @@ def all_gather_cross_replica_mean_grad(x: torch.Tensor, replication_factor: int)
...
+    """
+    All-gather across IPU program replicas.
+
+    Gathers and stacks tensors occupying the same memory location across all IPUs
+
+    x -- shape (*)
+    returns --  shape (replication_factor, *)
+    """
+    x = all_gather_cross_replica_identical_grads_in(x, replication_factor)
+    x = all_reduce_cross_replica_sum(x, replication_factor, True)
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/_impl/collectives.py]: Perhaps use a keyword arg `insert_in_grad_graph=True` just to make this very obvious to read.

File: tests/utils.py [hunk 2]
```diff
@@ -1,16 +1,22 @@
...
 import torch.nn as nn
 from poptorch.enums import CommGroupType, VariableRetrievalMode
 
 
 def _apply_replica_grouping(
-    model: nn.Module, comm_group_type: CommGroupType, shards: int
+    model: nn.Module,
+    comm_group_type: CommGroupType,
+    shards: int,
+    excluded_parameters: List[str] = [],
```
Comment (DouglasOrr) [inline tests/utils.py:13]: Maybe could add a `assert not (set(excluded_parameters) - set(dict(model.named_parameters())))` or something to make sure that there's no misspecified exclusions.

File: poptorch_experimental_addons/sharded.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    X_local_outer_dim, _ = X.shape
+    _, Y_local_outer_dim = Y.shape
+    result = torch.zeros((X_local_outer_dim, Y_local_outer_dim * replication_factor))
+    X = einops.rearrange(X, "m (k c) -> m k c", k=num_chunks)
+    Y = einops.rearrange(Y, "(k c) n -> c k n", k=num_chunks)
+    for i in range(num_chunks):
+        index = torch.tensor([i])
+        Yg = all_gather_cross_replica(
+            torch.index_select(Y, dim=1, index=index).squeeze(), replication_factor
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/sharded.py:40]: I thought you were having to use `index_select` for dynamic indices, but this is static. Out of curiosity, what happens if you do `Y[:, i]`?

File: poptorch_experimental_addons/sharded.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        )
+        Yg = einops.rearrange(Yg, "r c n -> c (n r)")
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/sharded.py]: I think I'd have expected to see `(r n)` here. So does this mean the column-sharding axis is "minor"?

File: poptorch_experimental_addons/sharded.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    X = all_reduce_cross_replica_sum(X, replication_factor, insert_in_grad_graph=True)
+    return X @ Y
+
+
+def colrowrep_sharded_matmul(
+    X: torch.Tensor, Y: torch.Tensor, replication_factor: int
+) -> Any:
+    """Matrix multiplation for row-sharded x column-sharded -> replicated tensors"""
+    out = X @ Y
+    return all_reduce_cross_replica_sum(out, replication_factor)
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/sharded.py:63]: So if I understand this correctly, this collective will only run in the forwards pass (I guess this is an existing-code naming issue that I failed to pick up on before, as I would expect `all_reduce_cross_replica_sum` to be "consistent" and run in both passes by default).

So this & `repcolcol_sharded_matmul` run under the assumption "replicated tensors have replicated gradients", so in this case there's no need to do it in the backwards pass. Is this right? In which case, worth a comment to that effect.

File: poptorch_experimental_addons/sharded.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
```
Comment (DouglasOrr) [inline poptorch_experimental_addons/sharded.py:2]: Worth putting a summary docstring here, I think.

File: tests/test_sharded_matmul.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    grad_Y = _sharding_transform_map[tester.Y_sharding]["out_test"](grad_Y, r=num_ipus)
+    return out, grad_X, grad_Y, tester.X.data, tester.Y.data
+
+
+@pytest.mark.parametrize("op", list(_op_mapping.keys()))
+def test_sharded_matmul(op: Callable[[Any], torch.Tensor]) -> None:
+    X, Y = generate_inputs()
+    num_ipus = 2
+    actual = run_sharded_matmul(X, Y, op, num_ipus)
+    expected = simulate_sharded_matmul(X, Y, op, num_ipus)
```
Comment (DouglasOrr) [inline tests/test_sharded_matmul.py:267]: I was going to suggest a plain matmul as reference, but now actually I wonder if it's useful to have both this & keeping the simulator. This would be pretty encouraging due to its simplicity.

**Review by DouglasOrr — COMMENTED** [review]
This looks good, just a few minor comments & questions (sorry if I keep asking the same questions—I think I'm being a bit dim about the semantics of our various collectives.) 

I'm afraid I have discovered serious review fatigue & simply cannot get my mind in gear w.r.t. the details of sharding, so I've reviewed all but the tests. @AlCatt91 can I rely on your review there?

File: poptorch_experimental_addons/sharded.py (hunk 3 above)
Comment (lyprince) [reply poptorch_experimental_addons/sharded.py:40]:
```
poptorch.poptorch_core.Error: In unknown:0: 'popart_exception': Failure in BasePadOp::runtimeConfirmShapes, for 390 (ai.graphcore.SliceGrad:1). This with pads of size 4, and input tensor of rank 1. Expected pads to be exactly 2x the unput Tensor's rank.
```

File: tests/utils.py (hunk 2 above)
Comment (lyprince) [reply tests/utils.py:13]: By "misspecified exclusions" do you mean `excluded_parameters` that are not in the named model parameters in the first place?

File: poptorch_experimental_addons/sharded.py (hunk 4 above)
Comment (lyprince) [reply poptorch_experimental_addons/sharded.py]: I believe so. I'm not confident on the major/minor terminology, but the way I think about this is by saying "Here's a list of `r` cxn matrices. Stack them along the column dimension". `(r n)` intuitively means to me "interleave them along the column dimension".

File: poptorch_experimental_addons/sharded.py (hunk 5 above)
Comment (lyprince) [reply poptorch_experimental_addons/sharded.py:63]: > I would expect all_reduce_cross_replica_sum to be "consistent" and run in both passes by default

Why would you expect this? The gradient of `all_reduce_cross_replica_sum` is a copy.

> So this & repcolcol_sharded_matmul run under the assumption "replicated tensors have replicated gradients", so in this case there's no need to do it in the backwards pass. Is this right? In which case, worth a comment to that effect.

Yes, I think this is correct.`repcolcol_sharded_matmul` and `rowcolrep_sharded_matmul` are functions to support the two types of matrix multiplication in the megatron-style tensor parallel case. The left hand block is `repcolcol` and the right hand block is `colrowrep`. The output of the grad op of `rowcolrep_sharded_matmul` should be row-sharded (so no all-reduce needed in the backward pass), and the output of the grad op of `repcolcol_sharded_matmul` should be replicated, (so all-reduce is needed in the backward pass).

Screenshot 2023-03-14 at 12 05 42

File: tests/test_sharded_matmul.py (hunk 7 above)
Comment (lyprince) [reply tests/test_sharded_matmul.py:267]: I did originally have this for forward pass only to check the simulated matmul produced the correct result (I'm not sure it would make sense with sharded/partial grads). I can bring this back in

**Comment by AlCatt91** [comment]
> I'm afraid I have discovered serious review fatigue & simply cannot get my mind in gear w.r.t. the details of sharding, so I've reviewed all but the tests. @AlCatt91 can I rely on your review there?

Yes of course, happy to take a closer look at the tests.

File: poptorch_experimental_addons/sharded.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import torch
+
+from .collectives import all_gather_cross_replica, all_reduce_cross_replica_sum
+
+
+def rowcolrow_sharded_matmul(
+    X: torch.Tensor, Y: torch.Tensor, replication_factor: int, num_chunks: int = 1
+) -> Any:
+    """
+    Matrix multiplation for row-sharded x column-sharded -> row-sharded tensors
```
Comment (AlCatt91) [inline poptorch_experimental_addons/sharded.py]: Typo (multiplation) - same in `repcolcol_sharded_matmul` and `colrowrep_sharded_matmul`

File: tests/test_sharded_matmul.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        "out_sim": lambda x: x,
+        "out_test": lambda x, r: x,
+    },
+    Sharding.Row: {
+        "in": partial(einops.rearrange, pattern="(r n) d -> r n d"),
+        "out_sim": partial(einops.rearrange, pattern="r n d -> (r n) d"),
+        "out_test": lambda x, r: x,
+    },
+    Sharding.Column: {
+        "in": partial(einops.rearrange, pattern="d (m r) -> r d m"),
```
Comment (AlCatt91) [inline tests/test_sharded_matmul.py]: I think Doug was right on this point, for column sharding the rearrange pattern I believe should be "d (r m) -> r d m" - otherwise the output is different from the plain matmul X @ Y for `colrowrep_sharded_matmul`. Right now this is compensating the similar patterns on Yg (both in `sharded.py` and in the simulator), which is why it's not throwing errors for `rowcolrow_sharded_matmul`. Inverting the pattern in all 3 points seem to be the correct way to have the check against X @ Y pass for all 3 ops.

File: tests/test_sharded_matmul.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+
+@pytest.mark.parametrize("op", list(_op_mapping.keys()))
+def test_sharded_matmul(op: Callable[[Any], torch.Tensor]) -> None:
+    X, Y = generate_inputs()
+    out = X @ Y
+    num_ipus = 2
+    actual = run_sharded_matmul(X, Y, op, num_ipus)
+    expected = simulate_sharded_matmul(X, Y, op, num_ipus)
+    assert_close(out, expected[0])
```
Comment (AlCatt91) [inline tests/test_sharded_matmul.py]: I am sure you are already on this, when checking against X @ Y need to discard half of the output for `colrowrep_sharded_matmul` and rearrange (hstack vs vstack) for output that is column sharded (`repcolcol_sharded_matmul`)

File: poptorch_experimental_addons/sharded.py (hunk 5 above)
Comment (AlCatt91) [reply poptorch_experimental_addons/sharded.py:63]: I believe the way it's implemented makes sense,

File: poptorch_experimental_addons/sharded.py (hunk 4 above)
Comment (AlCatt91) [reply poptorch_experimental_addons/sharded.py]: See comment below.

**Review by AlCatt91 — CHANGES_REQUESTED** [review]

File: tests/test_sharded_matmul.py (hunk 9 above)
Comment (lyprince) [reply tests/test_sharded_matmul.py]: You're right. Testing simulators against X@Y revealed these problems. Have corrected.

**Review by AlCatt91 — APPROVED** [review]

**Comment by AlCatt91** [comment]
LGTM

File: poptorch_experimental_addons/sharded.py (hunk 3 above)
Comment (DouglasOrr) [reply poptorch_experimental_addons/sharded.py:40]: Thanks for this—depressing, but good to know!
